Fix target-cpu fpu features on Armv8-R.

chrisnc opened the pull request:
This is a follow-up to #123159, but applied to Armv8-R.

This required https://github.com/llvm/llvm-project/pull/88287 to work properly. Now that this change exists in rustc's llvm, we can fix Armv8-R's default fpu features. In Armv8-R's case, the default features from LLVM for floating-point are sufficient, because there is no integer-only variant of this architecture.

chrisnc committed:
Fix target-cpu fpu features on Armv8-R.

This is a follow-up to #123159, but applied to Armv8-R.

This required https://github.com/llvm/llvm-project/pull/88287 to work
properly. Now that this change exists in rustc's llvm, we can fix
Armv8-R's default fpu features. In Armv8-R's case, the default features
from LLVM for floating-point are sufficient, because there is no
integer-only variant of this architecture.

jieyouxu commented:
And since it might help bisection in the future,
@bors rollup=never

chrisnc commented:
Confirmed that with this change, `-C target-cpu=cortex-r52` now allows double-precision and neon instructions to be generated, according to the additional features implied by this CPU in LLVM.

workingjubilee commented:
@chrisnc My apologies, for your good deeds will not go unpunished, as now I have to ask:

Can you add that as a test?

chrisnc commented:
Punishment accepted. Could you point me to an example that would be good to follow? The tests I see in `tests/assembly` seem to not have any operators, and adding the code like above in one of those tests requires turning on a bunch of lang items and I can't seem to get it to work...

Here's roughly where I got, but it doesn't compile because no_core disables everything, and if I try to use core, it won't be available in the test. I assume I'm doing this wrong.

```rust
// Test that Neon instructions are emitted when a target-cpu supporting them is specified.

//@ assembly-output: emit-asm
//@ compile-flags: --target armv8r-none-eabihf
//@ compile-flags: -C target-cpu=cortex-r52
//@ compile-flags: -C opt-level=3
//@ needs-llvm-components: arm

#![feature(no_core, lang_items)]
#![no_std]
#![no_core]
#![crate_type = "lib"]

#[lang = "sized"]
trait Sized {}

#[lang = "copy"]
trait Copy {}

// CHECK-LABEL: vadd_q:
#[no_mangle]
pub fn vadd_q(x: &mut [f32; 4], y: &[f32; 4]) {
    // CHECK: vld{{.*}}
    // CHECK: vld{{.*}}
    // CHECK: vadd.f32 {{.*}}
    // CHECK: vst{{.*}}
    // CHECK: bx lr
    let mut i = 0usize;
    while i < 4usize {
        x[i] += y[i];
        i += 1usize;
    }
}

// CHECK-LABEL: vadd_f64:
#[no_mangle]
pub fn vadd_f64(x: f64, y: f64) -> f64 {
    // CHECK: vadd.f64 d0, d0, d1
    // CHECK: bx lr
    x + y
}
```

workingjubilee commented:
The problem is that this situation is cross-compiling, but I don't believe we actually have a satisfactory way to run cross-compiled tests, so I think the actually most satisfactory way is unfortunately probably rmake and filecheck.

chrisnc commented:
Okay, the rmake test is working locally now at least. I'm not sure how to have the same test run for multiple targets, but I've set it up so it will extend to the other target+cpu combinations easily (tested `thumbv8m.main-none-eabihf` and `cortex-m85` also...).

workingjubilee commented:
no need to test every single possible subarchitecture, but yes, this looks nice.

I will see if we can't implement actual cross-compiled assembly testing, but for now, this will do quite well, thank you.

@bors r+ rollup

bors commented:
:pushpin: Commit d7b27907a644a95ababc9438882d8e7fcf386a6a has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee d7b27907a644a95ababc9438882d8e7fcf386a6a -->
<!-- homu: {"type":"Approved","sha":"d7b27907a644a95ababc9438882d8e7fcf386a6a","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

chrisnc commented:
Great! Will this actually run in CI? I had to provide a matching target with:
`./x test --target=armv8r-none-eabihf tests/run-make/arm-target-cpu-fp` for this to run properly. Is there somewhere I need to specify that?

chrisnc commented:
Sorry, needed to amend because I realized two things:
my .checks file had leading "//" which are superfluous.

I read my comments on the older PR and was reminded that the plan for Armv8-R is that the feature list should be empty, because Arm defines Armv8-R to always have these floating-point features, and so does LLVM, so we don't need to specify them. I've removed the feature setting and updated the comment and commit message to reflect that.

workingjubilee commented:
@bors r-

workingjubilee commented on tests/run-make/arm-target-cpu-fp/rmake.rs:
```diff
@@ -0,0 +1,37 @@
+    for case in CASES {
+        let lib = static_lib_name(case.cpu);
+
+        rustc()
+            .edition("2021")
+            .arg(format!("--target={}", case.target))
+            .arg(format!("-Ctarget-cpu={}", case.cpu))
+            .arg("-Copt-level=3")
+            .crate_type("rlib")
+            .input("lib.rs")
+            .output(&lib)
+            .run();
```
@chrisnc Each host that runs the run-make test suite, which includes all tier 1 hosts at least, should execute every rmake test.

Here, you explicitly built the code for each target we are checking cross-compilation against.

So yes, it should be run for each target.

Is there something that seems incomplete about this explanation?

workingjubilee replied:
Oh, I see, you're talking about the lack of precompiled stds.

workingjubilee replied:
hrmm, you raise a good point, this will work if we had a tier 2 target we could test this for, because then std will be compiled, but I can't see one...

chrisnc replied:
Yeah, I think the expectation is clear, I'm just not sure where to specify in CI that this test needs to be built with `--target=`. It didn't work without this on thumbv8m.main either, which is tier2. I needed to provide `./x test --target=...` in each case.

workingjubilee replied:
right. CI *will* build the stdlibs for these targets, but locally you won't.

can you set up a variant for thumbv8.main then?

chrisnc replied:
Aha. Will do. Is there a way to invoke the test such that this happens automatically?

chrisnc replied:
Added. If I include only one case and invoke it with the correct `--target=...`, each of them works. Still not sure how to have the entire test run with all of the cases in one go though.

chrisnc replied:
Oh, I missed your earlier comments about tier 2. I will remove the armv8r one. There are tests I can add for thumbv7em and armv7r though...

jieyouxu replied:
Can you leave a FIXME(#130375) for this? In the meantime, I think it's fine if it's somewhat limited for now.

chrisnc replied:
Added a FIXME to the top of rmake.rs, as well as expanding on what the test is for.

workingjubilee commented on tests/run-make/arm-target-cpu-fp/rmake.rs:
```diff
@@ -0,0 +1,37 @@
+            .run()
+            .stdout_utf8();
+
+        let check_file = format!("{}_{}.checks", case.target, case.cpu).replace("-", "_");
```
hmm, is there a reason we have to use `.replace` here, btw?

chrisnc replied:
I was following the convention that `.checks` files in other rmake tests have underscores, but the target and cpu names have dashes. Is there a better way to achieve this?

workingjubilee replied:
...I'm not sure why they do lmao.

workingjubilee replied:
oh there's only one rmake test with `.checks` files.

workingjubilee replied:
@chrisnc despite having multiple .checks files that's actually a convention of 1, it's not actually convened yet. :^)

chrisnc replied:
Ack. Hyphens it is.

chrisnc replied:
Updated. Just using the cpu name now, as that uniquely identifies the architecture anyway...

chrisnc commented:
Okay, I think it's in a good state now, pending CI... Thanks for all the feedback.

jieyouxu reviewed (approved):
Yes thanks, that looks good. The cross-compilation situation is indeed very unfortunate.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10867041962/job/30155358761) [(plain)](https://github.com/rust-lang/rust/commit/1df34e4a8230fb89963229346be4d7cab759b732/checks/30155358761/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:fb97a1d6377f6cf2227825318ca4bbde3889e0c420746f5a03ba46a07e9a862725c26a09b9fc49a0d129ebd75935d3f6cd19acf41cc4267a6846fd4aa574b12c:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.66s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/arm-target-cpu-features stdout ----

error: rmake recipe failed to complete
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out" && env -u RUSTFLAGS AR="ar" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-17/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerparallel dwp engine executionengine extensions filecheck frontendhlsl frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-17/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake"
--- stderr -------------------------------
command failed at line 46
command failed at line 46
Command { cmd: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out" "--edition" "2021" "--target=thumbv7em-none-eabihf" "-Copt-level=3" "--crate-type" "rlib" "lib.rs" "-o" "libcortex-m7.a", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc", defused: true, armed_location: Location { file: "/checkout/tests/run-make/arm-target-cpu-features/rmake.rs", line: 26, col: 27 } }, already_executed: true }
output status: `exit status: 1`
=== STDOUT ===


=== STDERR ===
error[E0463]: can't find crate for `core`
```

</details>

workingjubilee commented:
...huh, I guess that doesn't in fact build it? is it because it's not a dist build?

workingjubilee commented:
going to try a spread of dist builds to see if I was even half-right:

@bors try

bors commented:
:hourglass: Trying commit 1df34e4a8230fb89963229346be4d7cab759b732 with merge 1cdb483bf2c74fd5ba750ba79b01e2ceed12daa3...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1df34e4a8230fb89963229346be4d7cab759b732","merge_sha":"1cdb483bf2c74fd5ba750ba79b01e2ceed12daa3"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10867432428/job/30156785266)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10867432428/job/30156785266","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-various-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10867432428/job/30156186434) [(plain)](https://github.com/rust-lang-ci/rust/commit/1cdb483bf2c74fd5ba750ba79b01e2ceed12daa3/checks/30156186434/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
---- [run-make] tests/run-make/arm-target-cpu-features stdout ----

error: rmake recipe failed to complete
status: exit status: 1
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out" && env -u RUSTFLAGS AR="arm-none-eabi-ar" CC="arm-none-eabi-gcc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -mthumb -march=armv6s-m" CXX="arm-none-eabi-g++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -mthumb -march=armv6s-m" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_LINKER="arm-none-eabi-gcc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="thumbv6m-none-eabi" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/thumbv6m-none-eabi/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake"
--- stderr -------------------------------
command failed at line 46
command failed at line 46
Command { cmd: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/arm-target-cpu-features/rmake_out" "--edition" "2021" "--target=thumbv7em-none-eabihf" "-Copt-level=3" "--crate-type" "rlib" "lib.rs" "-o" "libcortex-m7.a", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc", defused: true, armed_location: Location { file: "/checkout/tests/run-make/arm-target-cpu-features/rmake.rs", line: 26, col: 27 } }, already_executed: true }
output status: `exit status: 1`
=== STDOUT ===


=== STDERR ===
error[E0463]: can't find crate for `core`
```

</details>

chrisnc commented:
:(
Is there a way to build core from within the rmake.rs so that this will work...?

workingjubilee commented:
> And would it be okay to break off the test part to another PR and merge the Armv8-R change as-is?

Yeah let's do that.

workingjubilee commented:
I guess we'll be stuck just straight-up invoking what amounts to `-Zbuild-std=core,compiler-builtins` for this.

chrisnc commented:
Done. Created https://github.com/rust-lang/rust/pull/130382 for the test.

workingjubilee commented:
coolio.

@bors r+ rollup

bors commented:
:pushpin: Commit 1a0ba0117794bb80e540bb449c4109b8f20060c1 has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 1a0ba0117794bb80e540bb449c4109b8f20060c1 -->
<!-- homu: {"type":"Approved","sha":"1a0ba0117794bb80e540bb449c4109b8f20060c1","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
> I guess we'll be stuck just straight-up invoking what amounts to `-Zbuild-std=core,compiler-builtins` for this.

These ones are really unfortunate, there's a couple of run-make tests that seem to need to use bootstrap cargo to `-Zbuild-std`, but those are really slow. Oh well.
